sections/generic-best-practices/test-code-with-debugging-in-mind.md

```diff
@@ -0,0 +1,345 @@
+  }`,
+  createMutationCustomType: `type LoginResponse {
+    accessToken: String!
+  }
+  `,
+  createMutationHandler: 'https://hasura-actions-demo.glitch.me/login',
+  // ...
+}
+```
+
+The rationale behind the short functions is creating small, reusable pieces of code to help other tests that must do similar things on the page.
+
+I think it is not good: because **it is hard to build a mental model of what the test does**! All the test parts are split into small functions and utilities, while the test's code must be as straightforward as possible.
+
+Can you recall the two underrated points at the top of the chapter? The idea is that the test should accomplish three main things:
+
+- To help the reader to understand what code does
+- To allow debugging itself with ease
+
+The former requires the test's code to be as dumb as possible, and there is no advantage in having abstraction in the test's code because it leads to spending more time debugging and maintaining the tests instead of the application.
+
+The latter relates to the wrong part of tests: debugging/fixing them. Debugging a UI test is hard because you work with
+
+- You front-end application
+- The browser
+- A tool that controls the browser
+- The instructions you provide to the tool that controls the browser
+
+Each of the above elements can fail, and even the more experienced developers can struggle with understanding the source of a failing test.
+
+Hence, E2E testing is complicated. Cypress improves the developers' life (find more in [Some UI testing problems and the Cypress way](../tools/ui-testing-problems-cypress.md) chapter), but a straightforward code dramatically helps.
+
+### No abstraction at all
+
+I recommend having no abstraction at all (later, I will talk about the exceptions and which abstraction it's good)! I rewrote the above example to something like
+
+```ts
+it('Test the feature', () => {
+  cy.get('textarea').eq(0).as('actionDefinitionTextarea');
+  cy.get('textarea').eq(1).as('typeConfigurationTextarea');
+
+  cy.get('@actionDefinitionTextarea').clearConsoleTextarea().type(
+    `type Mutation {
+        login (username: String!, password: String!): LoginResponse
+      }`,
+    { force: true, delay: 0 }
+  );
+
+  cy.get('@typeConfigurationTextarea').clearConsoleTextarea().type(
+    `type LoginResponse {
+      accessToken: String!
+    }
+    `,
+    { force: true, delay: 0 }
+  );
+
+  // ...
+})
+```
+
+The rewritten test does the same as the original one, but when you jump on the test's code, the overhead of jumping back and forth to connect the dots mentally is not needed.
+
+- Do you want to know that is typed in the textarea? No hassle, it's there!
+- Do you want to know what is the textareas used by the text? No hassle, it's there!
+
+### When is abstraction good in tests?
+
+In my opinion:
+
+- When I want to hide some test oddities that could distract the readers for no value
+- When they are soft, almost parameters-free, only one-level deep
+
```

Comment:
I would add here: when there is a considerable amount of duplication, 3+ references can justify some abstraction imo.

Reply:
Done and pushed (with a slight change, though, I prefer not to report "3", LMK WDYT)

Reply:
sounds good